src/Lean/Meta/Basic.lean

```diff
@@ -1506,40 +1518,48 @@ private def forallBoundedTelescopeImp (type : Expr) (maxFVars? : Option Nat) (k
+  process consumeLambda consumeLet (← getLCtx) #[] e
 where
-  process (consumeLet : Bool) (lctx : LocalContext) (fvars : Array Expr) (e : Expr) : MetaM α := do
-    match fvarsSizeLtMaxFVars fvars maxFVars?, consumeLet, e with
-    | true, _, .lam n d b bi =>
+  process (consumeLambda : Bool) (consumeLet : Bool) (lctx : LocalContext) (fvars : Array Expr) (e : Expr) : MetaM α := do
+    let finish (e : Expr) : MetaM α :=
+      let e := e.instantiateRevRange 0 fvars.size fvars
+      withReader (fun ctx => { ctx with lctx := lctx }) do
+        withNewLocalInstancesImp fvars 0 do
+          k fvars e
+    match fvarsSizeLtMaxFVars fvars maxFVars?, consumeLambda, consumeLet, e with
+    | true, true, _, .lam n d b bi =>
       let d := d.instantiateRevRange 0 fvars.size fvars
       let d := if cleanupAnnotations then d.cleanupAnnotations else d
       let fvarId ← mkFreshFVarId
       let lctx := lctx.mkLocalDecl fvarId n d bi
       let fvar := mkFVar fvarId
-      process consumeLet lctx (fvars.push fvar) b
-    | true, true, .letE n t v b _ => do
-      let t := t.instantiateRevRange 0 fvars.size fvars
-      let t := if cleanupAnnotations then t.cleanupAnnotations else t
-      let v := v.instantiateRevRange 0 fvars.size fvars
-      let fvarId ← mkFreshFVarId
-      let lctx := lctx.mkLetDecl fvarId n t v
-      let fvar := mkFVar fvarId
-      process true lctx (fvars.push fvar) b
-    | _, _, e =>
-      let e := e.instantiateRevRange 0 fvars.size fvars
-      withReader (fun ctx => { ctx with lctx := lctx }) do
-        withNewLocalInstancesImp fvars 0 do
-          k fvars e
+      process true consumeLet lctx (fvars.push fvar) b
+    | true, _, true, .letE n t v b nondep => do
+      if !nondep && nondepLetOnly then
+        finish e
+      else
+        let t := t.instantiateRevRange 0 fvars.size fvars
+        let t := if cleanupAnnotations then t.cleanupAnnotations else t
+        let v := v.instantiateRevRange 0 fvars.size fvars
+        let fvarId ← mkFreshFVarId
+        let lctx := lctx.mkLetDecl fvarId n t v (nondep && preserveNondepLet)
+        let fvar := mkFVar fvarId
+        process consumeLambda true lctx (fvars.push fvar) b
+    | _, _, _, e =>
+      finish e
 
 /--
 Similar to `lambdaTelescope` but for lambda and let expressions.
 
-If `cleanupAnnotations` is `true`, we apply `Expr.cleanupAnnotations` to each type in the telescope.
+- If `cleanupAnnotations` is `true`, we apply `Expr.cleanupAnnotations` to each type in the telescope.
+- If `preserveNondep` is `false`, all `have`s are converted to `let`s.
+
+See also `mapLambdaLetTelescope` for entering and rebuilding the telescope.
 -/
-def lambdaLetTelescope (e : Expr) (k : Array Expr → Expr → n α) (cleanupAnnotations := false) : n α :=
-  map2MetaM (fun k => lambdaTelescopeImp e true .none k (cleanupAnnotations := cleanupAnnotations)) k
+def lambdaLetTelescope (e : Expr) (k : Array Expr → Expr → n α) (cleanupAnnotations := false) (preserveNondepLet := true) : n α :=
```

Comment:
In the current `master`, `lambdaLetTelescope` stops at `have`. For example, the following example prints 1.
```lean
import Lean
def c : Nat → Nat → Bool := fun x => have y := 1; fun z => x == y + z
open Lean Meta in
run_meta do
  let decl ← getConstInfo ``c
  lambdaLetTelescope decl.value! fun xs _ => IO.println xs.size  
```
I am not defending the current behavior. 
The comment above suggests that this example will print `3` if `preserveNondepLet := false`. What happens if `preserveNondepLet := true`. I am assuming it will "consume" the `have` and mark it as `nondep` in the local context.

Reply:
Yes, now `have`s are always consumed and that prints `3`. Setting `preserveNondepLet` controls whether the nondep flag will be cleared. I've added this to the tests:
```lean
def c : Nat → Nat → Bool := fun x => have y := 1; fun z => x == y + z
/--
info: #[false, true, false]
#[false, false, false]
-/
#guard_msgs in
open Lean Meta in
run_meta do
  let decl ← getConstInfo ``c
  lambdaLetTelescope decl.value! fun xs _ => do
    IO.println <| ← xs.mapM fun x => return (← x.fvarId!.getDecl).isNondep
  lambdaLetTelescope decl.value! (preserveNondepLet := false) fun xs _ => do
    IO.println <| ← xs.mapM fun x => return (← x.fvarId!.getDecl).isNondep
```